sdk/core/Azure.Core/src/Shared/AsyncLockWithValue.cs

```diff
@@ -0,0 +1,230 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure.Core.Pipeline;
+
+namespace Azure.Core
+{
+    /// <summary>
+    /// Primitive that combines async lock and value cache
+    /// </summary>
+    /// <typeparam name="T"></typeparam>
+    internal sealed class AsyncLockWithValue<T>
+    {
+        private readonly object _syncObj = new object();
+        private Queue<TaskCompletionSource<Lock>>? _waiters;
+        private bool _isLocked;
+        private bool _hasValue;
+        private T? _value;
+
+        public bool HasValue
+        {
+            get
+            {
+                lock (_syncObj)
+                {
+                    return _hasValue;
+                }
+            }
+        }
+
+        public AsyncLockWithValue() { }
+
+        public AsyncLockWithValue(T value)
+        {
+            _hasValue = true;
+            _value = value;
+        }
+
+        public bool TryGetValue(out T? value)
+        {
+            lock (_syncObj)
```

Comment:
Any perf advantage to using a `ReaderWriterLockSlim` instead?

Reply:
Only for scenario when lots of subscribers decided to cancel their wait, and have done so concurrently, but that is an edge case. In more common scenarios overhead of the `ReaderWriterLockSlim` is higher than code bits in the critical section itself.